sql/parser/eval.go

```diff
@@ -535,6 +728,15 @@ var cmpOps = map[cmpArgs]func(Datum, Datum) (Datum, error){
 	cmpArgs{EQ, floatType, floatType}: func(left Datum, right Datum) (Datum, error) {
 		return DBool(left.(DFloat) == right.(DFloat)), nil
 	},
+	cmpArgs{EQ, dateType, dateType}: func(left Datum, right Datum) (Datum, error) {
+		return DBool(left.(DDate).Equal(right.(DDate).Time)), nil
+	},
+	cmpArgs{EQ, timestampType, timestampType}: func(left Datum, right Datum) (Datum, error) {
```

Comment:
The timezone offset is only used for parsing and presentation. In principle they should not be a part of the index, but we store entire values in the index to prevent not needing to store them outside of the index. I agree that we can have duplicates because of the time zone. We can document this as a shortcoming of using a timestamp as a unique index.